contracts/Logbook.sol

```diff
@@ -0,0 +1,79 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.0;
+
+import "./BatchNFT.sol";
+
+/**
+ * @dev Contract that allows token owner to read and write logbook
+ */
+abstract contract Logbook is BatchNFT {
+    struct Log {
+        address sender;
+        string message;
+        uint256 createdAt;
+    }
+    struct TokenLogbook {
+        Log[] logs;
```

Comment:
Should we use mapping instead of array here since mapping is cheaper and more scalable than array? (See [discussion](https://ethereum.stackexchange.com/questions/2592/store-data-in-mapping-vs-array), and seems also the reason @Carolusian designed [this data structure](https://github.com/thematters/matters-avatar/pull/7/files#diff-8cfec1bf9618127beb4f10dc2d78fa78d854a2b5507a53c8f49b85b758777b46R17-R19))

Reply:
```
struct TokenLogbookMapping {
    mapping(uint256 => Log) logs;
    uint256 count;
    bool isLocked;
}

mapping(uint256 => TokenLogbookMapping) private _logbookMapping;

function appendMappingLog(uint256 tokenId, string calldata message) public {
    require(
        _isApprovedOrOwner(_msgSender(), tokenId),
        "caller is not owner nor approved"
    );

    TokenLogbookMapping storage logbookMapping = _logbookMapping[tokenId];

    require(!logbookMapping.isLocked, "logbook is locked");
    require(
        bytes(message).length <= MAX_LOG_LENGTH,
        "log exceeds max length"
    );

    address owner = ERC721.ownerOf(tokenId);
    Log memory newLog = Log({
        sender: owner,
        message: message,
        createdAt: block.timestamp
    });

    logbookMapping.logs[logbookMapping.count + 1] = newLog;
    logbookMapping.count += 1;
    logbookMapping.isLocked = true;

    emit LogbookNewLog(tokenId, owner);
}
```

```
Gas used for appendLog (array, 1): 206647. Estimated ETH: 0.0206647
Gas used for appendLog (array, 2): 189547. Estimated ETH: 0.0189547
Gas used for appendLog (array, 3): 189547. Estimated ETH: 0.0189547
Gas used for appendLog (array, 4): 189547. Estimated ETH: 0.0189547
Gas used for appendLog (array, 5): 189547. Estimated ETH: 0.0189547
Gas used for appendLog (array, 6): 189547. Estimated ETH: 0.0189547
Gas used for appendLog (array, 7): 189547. Estimated ETH: 0.0189547
Gas used for appendLog (array, 8): 189547. Estimated ETH: 0.0189547
Gas used for appendLog (array, 9): 189547. Estimated ETH: 0.0189547
Gas used for appendLog (array, 10): 189547. Estimated ETH: 0.0189547
Gas used for appendLog (mapping, 1): 206998. Estimated ETH: 0.0206998
Gas used for appendLog (mapping, 2): 189898. Estimated ETH: 0.0189898
Gas used for appendLog (mapping, 3): 189898. Estimated ETH: 0.0189898
Gas used for appendLog (mapping, 4): 189898. Estimated ETH: 0.0189898
Gas used for appendLog (mapping, 5): 189898. Estimated ETH: 0.0189898
Gas used for appendLog (mapping, 6): 189898. Estimated ETH: 0.0189898
Gas used for appendLog (mapping, 7): 189898. Estimated ETH: 0.0189898
Gas used for appendLog (mapping, 8): 189898. Estimated ETH: 0.0189898
Gas used for appendLog (mapping, 9): 189898. Estimated ETH: 0.0189898
Gas used for appendLog (mapping, 10): 189898. Estimated ETH: 0.0189898
```

made a simple test for mapping, the gas cost on.`appendLog` is pretty similar to array.